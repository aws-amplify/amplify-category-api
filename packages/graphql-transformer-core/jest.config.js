const baseConfig = require('../../jest.config.base.js'); // eslint-disable-line @typescript-eslint/no-var-requires

module.exports = {
  ...baseConfig,
  coverageThreshold: {
    global: {
      branches: 65,
      functions: 33,
      lines: 47,
    },
  },
};
