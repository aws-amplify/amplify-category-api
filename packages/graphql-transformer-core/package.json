{
  "name": "graphql-transformer-core",
  "version": "8.2.4",
  "description": "A framework to transform from GraphQL SDL to AWS cloudFormation.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-category-api.git",
    "directory": "packages/graphql-transformer-core"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "cloudformation",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "build": "rimraf ./lib ./tsconfig.tsbuildinfo && tsc",
    "watch": "tsc -w",
    "clean": "rimraf ./lib",
    "extract-api": "ts-node ../../scripts/extract-api.ts"
  },
  "dependencies": {
    "cloudform-types": "^4.2.0",
    "deep-diff": "^1.0.2",
    "fs-extra": "^8.1.0",
    "glob": "^10.3.0",
    "graphql": "^15.5.0",
    "graphql-transformer-common": "4.27.1",
    "lodash": "^4.17.21"
  },
  "devDependencies": {
    "@aws-amplify/amplify-cli-core": "^4.2.10",
    "@types/fs-extra": "^8.0.1",
    "@types/node": "^18.17.0",
    "mock-fs": "5.2.0",
    "@types/mock-fs": "4.13.4"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true,
    "coverageProvider": "v8",
    "coverageThreshold": {
      "global": {
        "branches": 67,
        "functions": 33,
        "lines": 48
      }
    },
    "coverageReporters": [
      "clover",
      "text"
    ]
  }
}
