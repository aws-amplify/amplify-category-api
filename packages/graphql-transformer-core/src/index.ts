import './polyfills/Object.assign';
import { print } from 'graphql';
import { DeploymentResources } from './DeploymentResources';
import { TransformerContext, MappingParameters } from './TransformerContext';
import { Transformer } from './Transformer';
import { ITransformer } from './ITransformer';
import { GraphQLTransform } from './GraphQLTransform';
import { collectDirectiveNames, collectDirectivesByType, collectDirectivesByTypeNames } from './collectDirectives';
import { stripDirectives } from './stripDirectives';
import {
  buildProject as buildAPIProject,
  uploadDeployment as uploadAPIProject,
  migrateAPIProject,
  revertAPIMigration,
  CLOUDFORMATION_FILE_NAME,
  PARAMETERS_FILE_NAME,
} from './util/amplifyUtils';
import {
  readSchema as readProjectSchema,
  loadProject as readProjectConfiguration,
  loadConfig as readTransformerConfiguration,
  writeConfig as writeTransformerConfiguration,
} from './util/transformConfig';
import { EXTRA_DIRECTIVES_DOCUMENT } from './validation';

export * from './errors';
export * from './util';
export { getTableNameForModel } from './tableNameMap';

/**
 * Returns the set of directives that are supported by AppSync service
 */
export function getAppSyncServiceExtraDirectives(): string {
  return print(EXTRA_DIRECTIVES_DOCUMENT);
}
export { FeatureFlagProvider } from './FeatureFlags';

export {
  GraphQLTransform,
  TransformerContext,
  MappingParameters,
  Transformer,
  ITransformer,
  collectDirectiveNames,
  collectDirectivesByType,
  collectDirectivesByTypeNames,
  stripDirectives,
  buildAPIProject,
  migrateAPIProject,
  uploadAPIProject,
  readProjectSchema,
  readProjectConfiguration,
  readTransformerConfiguration,
  writeTransformerConfiguration,
  revertAPIMigration,
  DeploymentResources,
  CLOUDFORMATION_FILE_NAME,
  PARAMETERS_FILE_NAME,
};
// No-op change to trigger publish
