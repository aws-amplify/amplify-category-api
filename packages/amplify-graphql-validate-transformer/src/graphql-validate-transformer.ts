import { ValidateDirective } from '@aws-amplify/graphql-directives';
import { DirectiveWrapper, MappingTemplate, TransformerPluginBase } from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerSchemaVisitStepContextProvider,
  TransformerPluginProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { DirectiveNode, FieldDefinitionNode, InterfaceTypeDefinitionNode, ObjectTypeDefinitionNode } from 'graphql';
import { ValidateDirectiveArgs, ValidateDirectiveConfiguration, ValidationsByField, ValidationType } from './types';
import { validate } from './validators';
import { generateTypeValidationSnippet } from './vtl-generator';

export class ValidateTransformer extends TransformerPluginBase implements TransformerPluginProvider {
  private directiveMap = new Map<string, ValidateDirectiveConfiguration[]>();

  constructor() {
    super('amplify-graphql-validate-transformer', ValidateDirective.definition);
  }

  /**
   * Processes `@validate` directives and validates their usage and inputs.
   * @param parent - The parent object or interface type definition node containing the field
   * @param definition - The field definition node being processed
   * @param directive - The `@validate` directive node applied to the field
   * @param _ - The transformer schema visit step context provider (unused)
   */
  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    _: TransformerSchemaVisitStepContextProvider,
  ): void => {
    const directiveWrapped = new DirectiveWrapper(directive);
    const config = this.getValidateDirectiveConfiguration(directiveWrapped, parent as ObjectTypeDefinitionNode, definition);

    validate(parent as ObjectTypeDefinitionNode, definition, directive, config);

    const parentName = parent.name.value;
    if (!this.directiveMap.has(parentName)) {
      this.directiveMap.set(parentName, []);
    }
    this.directiveMap.get(parentName)!.push(config);
  };

  /**
   * Generates resolvers for validation directives.
   * @param ctx - The transformer context provider
   */
  generateResolvers = (ctx: TransformerContextProvider): void => {
    const mutationTypeName = ctx.output.getMutationTypeName();
    if (!mutationTypeName) {
      return;
    }

    // Generate a single validation snippet for each type
    for (const [typeName, configs] of this.directiveMap.entries()) {
      // Group validations by field
      const validationsByField: ValidationsByField = {};
      for (const config of configs) {
        const fieldName = config.fieldNode.name.value;
        if (!validationsByField[fieldName]) {
          validationsByField[fieldName] = [];
        }

        const defaultErrorMessage = constructDefaultErrorMessage(typeName, fieldName, config.validationType, config.validationValue);

        validationsByField[fieldName].push({
          validationType: config.validationType,
          validationValue: config.validationValue,
          errorMessage: config.errorMessage || defaultErrorMessage,
        });
      }

      // Generate a single combined VTL snippet for all validations in this type
      const combinedTypeSnippet = generateTypeValidationSnippet(typeName, validationsByField);

      // Add the combined validation to create mutation
      const createFieldName = `create${typeName}`;
      const createResolver = ctx.resolvers.getResolver(mutationTypeName, createFieldName);
      if (createResolver) {
        createResolver.addVtlFunctionToSlot(
          'validate',
          MappingTemplate.s3MappingTemplateFromString(
            combinedTypeSnippet,
            `${mutationTypeName}.${createFieldName}.{slotName}.{slotIndex}.req.vtl`,
          ),
        );
      }

      // Add the combined validation to update mutation
      const updateFieldName = `update${typeName}`;
      const updateResolver = ctx.resolvers.getResolver(mutationTypeName, updateFieldName);
      if (updateResolver) {
        updateResolver.addVtlFunctionToSlot(
          'validate',
          MappingTemplate.s3MappingTemplateFromString(
            combinedTypeSnippet,
            `${mutationTypeName}.${updateFieldName}.{slotName}.{slotIndex}.req.vtl`,
          ),
        );
      }
    }
  };

  /**
   * Extracts and formats the validation directive configuration from a `@validate` directive.
   * @param directive - The wrapped directive containing validation arguments
   * @param parentNode - The object type definition node that contains the field
   * @param fieldNode - The field definition node that the directive is applied to
   * @returns A configuration object containing the validation rules and metadata
   * @private
   */
  private getValidateDirectiveConfiguration(
    directive: DirectiveWrapper,
    parentNode: ObjectTypeDefinitionNode,
    fieldNode: FieldDefinitionNode,
  ): ValidateDirectiveConfiguration {
    const defaultArgs: ValidateDirectiveArgs = {
      type: '' as ValidationType,
      value: '',
      errorMessage: '',
    };
    const args = directive.getArguments<ValidateDirectiveArgs>(defaultArgs);

    return {
      parentNode,
      fieldNode,
      validationType: args.type,
      validationValue: args.value,
      errorMessage: args.errorMessage,
    };
  }
}

/**
 * Constructs a default error message based on the validation type and value.
 * @param typeName - The name of the type containing the field
 * @param fieldName - The name of the field being validated
 * @param validationType - The type of validation (e.g., gt, lt, minLength, etc.)
 * @param validationValue - The value used for validation
 * @returns A default error message string
 */
const constructDefaultErrorMessage = (
  typeName: string,
  fieldName: string,
  validationType: ValidationType,
  validationValue: string,
): string => {
  switch (validationType) {
    case 'gt':
      return `Field ${fieldName} of type ${typeName} must be greater than ${validationValue}`;
    case 'lt':
      return `Field ${fieldName} of type ${typeName} must be less than ${validationValue}`;
    case 'gte':
      return `Field ${fieldName} of type ${typeName} must be greater than or equal to ${validationValue}`;
    case 'lte':
      return `Field ${fieldName} of type ${typeName} must be less than or equal to ${validationValue}`;
    case 'minLength':
      return `Field ${fieldName} of type ${typeName} must have a minimum length of ${validationValue}`;
    case 'maxLength':
      return `Field ${fieldName} of type ${typeName} must have a maximum length of ${validationValue}`;
    case 'startsWith':
      return `Field ${fieldName} of type ${typeName} must start with ${validationValue}`;
    case 'endsWith':
      return `Field ${fieldName} of type ${typeName} must end with ${validationValue}`;
    case 'matches':
      return `Field ${fieldName} of type ${typeName} must match ${validationValue}`;
    default:
      throw new Error(`Unsupported validation type: ${validationType}`);
  }
};
