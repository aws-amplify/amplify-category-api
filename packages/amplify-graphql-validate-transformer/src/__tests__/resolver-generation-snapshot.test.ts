import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { validateModelSchema } from '@aws-amplify/graphql-transformer-core';
import { parse } from 'graphql';
import { testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { ValidateTransformer } from '../graphql-validate-transformer';

/**
 * Tests for the ValidateTransformer's resolver generation functionality.
 *
 * This test suite verifies that the transformer:
 * - Correctly generates VTL templates for validation rules
 * - Maintains consistent snapshot output for different validation scenarios
 * - Handles all supported validation types (numeric, string) correctly
 */
describe('Resolver generation', () => {
  const transformSchema = (schema: string): any => {
    const out = testTransform({
      schema,
      transformers: [new ModelTransformer(), new ValidateTransformer()],
    });
    expect(out).toBeDefined();
    validateModelSchema(parse(out.schema));
    return out;
  };

  const testValidation = (
    out: any,
    type: string,
    field: string,
    validationType: string,
    index: number,
    operation: 'create' | 'update',
  ): void => {
    expect(out.resolvers[`Mutation.${operation}${type}.validate.${index}.req.vtl`]).toMatchSnapshot(
      `${operation} ${type.toLowerCase()} ${field} ${validationType} validation`,
    );
  };

  const testValidationPair = (out: any, type: string, field: string, validationType: string, index: number): void => {
    testValidation(out, type, field, validationType, index, 'create');
    testValidation(out, type, field, validationType, index, 'update');
  };

  it('should generate correct validation resolvers for numeric validations', () => {
    const schema = `
      type Product @model {
        id: ID!
        price: Float! @validate(type: gt, value: "0", errorMessage: "Price must be positive")
          @validate(type: lt, value: "1000000", errorMessage: "Price must be less than 1,000,000")
        quantity: Int! @validate(type: gte, value: "0", errorMessage: "Quantity cannot be negative")
          @validate(type: lte, value: "100", errorMessage: "Quantity cannot exceed 100")
      }`;

    const out = transformSchema(schema);

    // Test price validations
    testValidationPair(out, 'Product', 'price', 'gt', 1);
    testValidationPair(out, 'Product', 'price', 'lt', 2);

    // Test quantity validations
    testValidationPair(out, 'Product', 'quantity', 'gte', 3);
    testValidationPair(out, 'Product', 'quantity', 'lte', 4);
  });

  it('should generate correct validation resolvers for string validations', () => {
    const schema = `
      type User @model {
        id: ID!
        email: String! @validate(type: matches, value: "^[A-Za-z0-9+_.-]+@(.+)$", errorMessage: "Invalid email format")
        username: String! @validate(type: minLength, value: "3", errorMessage: "Username too short")
          @validate(type: maxLength, value: "20", errorMessage: "Username too long")
        url: String @validate(type: startsWith, value: "https://", errorMessage: "URL must start with https://")
          @validate(type: endsWith, value: ".com", errorMessage: "URL must end with .com")
      }`;

    const out = transformSchema(schema);

    // Test email validation
    testValidationPair(out, 'User', 'email', 'matches', 1);

    // Test username validations
    testValidationPair(out, 'User', 'username', 'minLength', 2);
    testValidationPair(out, 'User', 'username', 'maxLength', 3);

    // Test url validations
    testValidationPair(out, 'User', 'url', 'startsWith', 4);
    testValidationPair(out, 'User', 'url', 'endsWith', 5);
  });

  it('should generate correct validation resolvers for multiple validations on the same field', () => {
    const schema = `
      type Post @model {
        id: ID!
        title: String! @validate(type: minLength, value: "5", errorMessage: "Title too short")
          @validate(type: maxLength, value: "100", errorMessage: "Title too long")
          @validate(type: startsWith, value: "prefix", errorMessage: "Title must start with prefix")
          @validate(type: endsWith, value: "suffix", errorMessage: "Title must end with suffix")
          @validate(type: matches, value: "^[A-Za-z0-9 ]+$", errorMessage: "Title can only contain letters, numbers and spaces")
      }`;

    const out = transformSchema(schema);

    // Test all title validations
    testValidationPair(out, 'Post', 'title', 'minLength', 1);
    testValidationPair(out, 'Post', 'title', 'maxLength', 2);
    testValidationPair(out, 'Post', 'title', 'startsWith', 3);
    testValidationPair(out, 'Post', 'title', 'endsWith', 4);
    testValidationPair(out, 'Post', 'title', 'matches', 5);
  });

  it('should generate correct validation resolvers for multiple fields', () => {
    const schema = `
      type Comment @model {
        id: ID!
        content: String! @validate(type: minLength, value: "1", errorMessage: "Content cannot be empty")
          @validate(type: maxLength, value: "1000", errorMessage: "Content too long")
        author: String! @validate(type: minLength, value: "3", errorMessage: "Author too short")
          @validate(type: maxLength, value: "20", errorMessage: "Author too long")
        rating: Int! @validate(type: gte, value: "0", errorMessage: "Rating cannot be negative")
          @validate(type: lte, value: "5", errorMessage: "Rating cannot exceed 5")
      }`;

    const out = transformSchema(schema);

    // Test content validations
    testValidationPair(out, 'Comment', 'content', 'minLength', 1);
    testValidationPair(out, 'Comment', 'content', 'maxLength', 2);

    // Test author validations
    testValidationPair(out, 'Comment', 'author', 'minLength', 3);
    testValidationPair(out, 'Comment', 'author', 'maxLength', 4);

    // Test rating validations
    testValidationPair(out, 'Comment', 'rating', 'gte', 5);
    testValidationPair(out, 'Comment', 'rating', 'lte', 6);
  });
});
