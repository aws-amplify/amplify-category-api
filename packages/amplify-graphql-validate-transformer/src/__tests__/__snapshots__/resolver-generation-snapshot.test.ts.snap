// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolver Generation Snapshot should generate correct combined validation resolver for multiple fields: create comment combined validations 1`] = `
"## [Start] Validations for type Comment. **

## [Start] Validations for field content. **
#if(!$util.isNull($ctx.args.input.content) )
  #set($minLengthValidationPassed = $ctx.args.input.content.length() >= 1)
  #if(!$minLengthValidationPassed)
    $util.error('Content cannot be empty')
  #end
  #set($maxLengthValidationPassed = $ctx.args.input.content.length() <= 1000)
  #if(!$maxLengthValidationPassed)
    $util.error('Content too long')
  #end
#end
## [End] Validations for field content. **

## [Start] Validations for field author. **
#if(!$util.isNull($ctx.args.input.author) )
  #set($minLengthValidationPassed = $ctx.args.input.author.length() >= 3)
  #if(!$minLengthValidationPassed)
    $util.error('Author too short')
  #end
  #set($maxLengthValidationPassed = $ctx.args.input.author.length() <= 20)
  #if(!$maxLengthValidationPassed)
    $util.error('Author too long')
  #end
#end
## [End] Validations for field author. **

## [Start] Validations for field rating. **
#if(!$util.isNull($ctx.args.input.rating) )
  #set($gteValidationPassed = $ctx.args.input.rating >= 0)
  #if(!$gteValidationPassed)
    $util.error('Rating cannot be negative')
  #end
  #set($lteValidationPassed = $ctx.args.input.rating <= 5)
  #if(!$lteValidationPassed)
    $util.error('Rating cannot exceed 5')
  #end
#end
## [End] Validations for field rating. **

## [End] Validations for type Comment. **

$util.toJson({})"
`;

exports[`Resolver Generation Snapshot should generate correct combined validation resolver for multiple fields: update comment combined validations 1`] = `
"## [Start] Validations for type Comment. **

## [Start] Validations for field content. **
#if(!$util.isNull($ctx.args.input.content) )
  #set($minLengthValidationPassed = $ctx.args.input.content.length() >= 1)
  #if(!$minLengthValidationPassed)
    $util.error('Content cannot be empty')
  #end
  #set($maxLengthValidationPassed = $ctx.args.input.content.length() <= 1000)
  #if(!$maxLengthValidationPassed)
    $util.error('Content too long')
  #end
#end
## [End] Validations for field content. **

## [Start] Validations for field author. **
#if(!$util.isNull($ctx.args.input.author) )
  #set($minLengthValidationPassed = $ctx.args.input.author.length() >= 3)
  #if(!$minLengthValidationPassed)
    $util.error('Author too short')
  #end
  #set($maxLengthValidationPassed = $ctx.args.input.author.length() <= 20)
  #if(!$maxLengthValidationPassed)
    $util.error('Author too long')
  #end
#end
## [End] Validations for field author. **

## [Start] Validations for field rating. **
#if(!$util.isNull($ctx.args.input.rating) )
  #set($gteValidationPassed = $ctx.args.input.rating >= 0)
  #if(!$gteValidationPassed)
    $util.error('Rating cannot be negative')
  #end
  #set($lteValidationPassed = $ctx.args.input.rating <= 5)
  #if(!$lteValidationPassed)
    $util.error('Rating cannot exceed 5')
  #end
#end
## [End] Validations for field rating. **

## [End] Validations for type Comment. **

$util.toJson({})"
`;

exports[`Resolver Generation Snapshot should generate correct combined validation resolver for multiple validations on the same field: create post combined validations 1`] = `
"## [Start] Validations for type Post. **

## [Start] Validations for field title. **
#if(!$util.isNull($ctx.args.input.title) )
  #set($minLengthValidationPassed = $ctx.args.input.title.length() >= 5)
  #if(!$minLengthValidationPassed)
    $util.error('Title too short')
  #end
  #set($maxLengthValidationPassed = $ctx.args.input.title.length() <= 100)
  #if(!$maxLengthValidationPassed)
    $util.error('Title too long')
  #end
  #set($startsWithValidationPassed = $ctx.args.input.title.startsWith('prefix'))
  #if(!$startsWithValidationPassed)
    $util.error('Title must start with prefix')
  #end
  #set($endsWithValidationPassed = $ctx.args.input.title.endsWith('suffix'))
  #if(!$endsWithValidationPassed)
    $util.error('Title must end with suffix')
  #end
  #set($matchesValidationPassed = $util.matches('^[A-Za-z0-9 ]+$', $ctx.args.input.title))
  #if(!$matchesValidationPassed)
    $util.error('Title can only contain letters, numbers and spaces')
  #end
#end
## [End] Validations for field title. **

## [End] Validations for type Post. **

$util.toJson({})"
`;

exports[`Resolver Generation Snapshot should generate correct combined validation resolver for multiple validations on the same field: update post combined validations 1`] = `
"## [Start] Validations for type Post. **

## [Start] Validations for field title. **
#if(!$util.isNull($ctx.args.input.title) )
  #set($minLengthValidationPassed = $ctx.args.input.title.length() >= 5)
  #if(!$minLengthValidationPassed)
    $util.error('Title too short')
  #end
  #set($maxLengthValidationPassed = $ctx.args.input.title.length() <= 100)
  #if(!$maxLengthValidationPassed)
    $util.error('Title too long')
  #end
  #set($startsWithValidationPassed = $ctx.args.input.title.startsWith('prefix'))
  #if(!$startsWithValidationPassed)
    $util.error('Title must start with prefix')
  #end
  #set($endsWithValidationPassed = $ctx.args.input.title.endsWith('suffix'))
  #if(!$endsWithValidationPassed)
    $util.error('Title must end with suffix')
  #end
  #set($matchesValidationPassed = $util.matches('^[A-Za-z0-9 ]+$', $ctx.args.input.title))
  #if(!$matchesValidationPassed)
    $util.error('Title can only contain letters, numbers and spaces')
  #end
#end
## [End] Validations for field title. **

## [End] Validations for type Post. **

$util.toJson({})"
`;

exports[`Resolver Generation Snapshot should generate correct combined validation resolver for numeric validations: create product combined validations 1`] = `
"## [Start] Validations for type Product. **

## [Start] Validations for field price. **
#if(!$util.isNull($ctx.args.input.price) )
  #set($gtValidationPassed = $ctx.args.input.price > 0)
  #if(!$gtValidationPassed)
    $util.error('Price must be positive')
  #end
  #set($ltValidationPassed = $ctx.args.input.price < 1000000)
  #if(!$ltValidationPassed)
    $util.error('Price must be less than 1,000,000')
  #end
#end
## [End] Validations for field price. **

## [Start] Validations for field quantity. **
#if(!$util.isNull($ctx.args.input.quantity) )
  #set($gteValidationPassed = $ctx.args.input.quantity >= 0)
  #if(!$gteValidationPassed)
    $util.error('Quantity cannot be negative')
  #end
  #set($lteValidationPassed = $ctx.args.input.quantity <= 100)
  #if(!$lteValidationPassed)
    $util.error('Quantity cannot exceed 100')
  #end
#end
## [End] Validations for field quantity. **

## [End] Validations for type Product. **

$util.toJson({})"
`;

exports[`Resolver Generation Snapshot should generate correct combined validation resolver for numeric validations: update product combined validations 1`] = `
"## [Start] Validations for type Product. **

## [Start] Validations for field price. **
#if(!$util.isNull($ctx.args.input.price) )
  #set($gtValidationPassed = $ctx.args.input.price > 0)
  #if(!$gtValidationPassed)
    $util.error('Price must be positive')
  #end
  #set($ltValidationPassed = $ctx.args.input.price < 1000000)
  #if(!$ltValidationPassed)
    $util.error('Price must be less than 1,000,000')
  #end
#end
## [End] Validations for field price. **

## [Start] Validations for field quantity. **
#if(!$util.isNull($ctx.args.input.quantity) )
  #set($gteValidationPassed = $ctx.args.input.quantity >= 0)
  #if(!$gteValidationPassed)
    $util.error('Quantity cannot be negative')
  #end
  #set($lteValidationPassed = $ctx.args.input.quantity <= 100)
  #if(!$lteValidationPassed)
    $util.error('Quantity cannot exceed 100')
  #end
#end
## [End] Validations for field quantity. **

## [End] Validations for type Product. **

$util.toJson({})"
`;

exports[`Resolver Generation Snapshot should generate correct combined validation resolver for string validations: create user combined validations 1`] = `
"## [Start] Validations for type User. **

## [Start] Validations for field email. **
#if(!$util.isNull($ctx.args.input.email) )
  #set($matchesValidationPassed = $util.matches('^[A-Za-z0-9+_.-]+@(.+)$', $ctx.args.input.email))
  #if(!$matchesValidationPassed)
    $util.error('Invalid email format')
  #end
#end
## [End] Validations for field email. **

## [Start] Validations for field username. **
#if(!$util.isNull($ctx.args.input.username) )
  #set($minLengthValidationPassed = $ctx.args.input.username.length() >= 3)
  #if(!$minLengthValidationPassed)
    $util.error('Username too short')
  #end
  #set($maxLengthValidationPassed = $ctx.args.input.username.length() <= 20)
  #if(!$maxLengthValidationPassed)
    $util.error('Username too long')
  #end
#end
## [End] Validations for field username. **

## [Start] Validations for field url. **
#if(!$util.isNull($ctx.args.input.url) )
  #set($startsWithValidationPassed = $ctx.args.input.url.startsWith('https://'))
  #if(!$startsWithValidationPassed)
    $util.error('URL must start with https://')
  #end
  #set($endsWithValidationPassed = $ctx.args.input.url.endsWith('.com'))
  #if(!$endsWithValidationPassed)
    $util.error('URL must end with .com')
  #end
#end
## [End] Validations for field url. **

## [End] Validations for type User. **

$util.toJson({})"
`;

exports[`Resolver Generation Snapshot should generate correct combined validation resolver for string validations: update user combined validations 1`] = `
"## [Start] Validations for type User. **

## [Start] Validations for field email. **
#if(!$util.isNull($ctx.args.input.email) )
  #set($matchesValidationPassed = $util.matches('^[A-Za-z0-9+_.-]+@(.+)$', $ctx.args.input.email))
  #if(!$matchesValidationPassed)
    $util.error('Invalid email format')
  #end
#end
## [End] Validations for field email. **

## [Start] Validations for field username. **
#if(!$util.isNull($ctx.args.input.username) )
  #set($minLengthValidationPassed = $ctx.args.input.username.length() >= 3)
  #if(!$minLengthValidationPassed)
    $util.error('Username too short')
  #end
  #set($maxLengthValidationPassed = $ctx.args.input.username.length() <= 20)
  #if(!$maxLengthValidationPassed)
    $util.error('Username too long')
  #end
#end
## [End] Validations for field username. **

## [Start] Validations for field url. **
#if(!$util.isNull($ctx.args.input.url) )
  #set($startsWithValidationPassed = $ctx.args.input.url.startsWith('https://'))
  #if(!$startsWithValidationPassed)
    $util.error('URL must start with https://')
  #end
  #set($endsWithValidationPassed = $ctx.args.input.url.endsWith('.com'))
  #if(!$endsWithValidationPassed)
    $util.error('URL must end with .com')
  #end
#end
## [End] Validations for field url. **

## [End] Validations for type User. **

$util.toJson({})"
`;
